gnd/src/main/java/com/google/android/gnd/ui/common/FeatureHelper.java

```diff
@@ -16,46 +16,73 @@
-import dagger.hilt.android.qualifiers.ApplicationContext;
+import com.google.android.gnd.model.feature.GeoJsonFeature;
 import dagger.hilt.android.scopes.ActivityScoped;
 import java8.util.Optional;
 import javax.inject.Inject;
+import org.jetbrains.annotations.Nullable;
 
 /** Common logic for formatting attributes of {@link Feature} for display to the user. */
 @ActivityScoped
 public class FeatureHelper {
 
-  private final Context context;
+  private final Resources resources;
 
   @Inject
-  FeatureHelper(@ApplicationContext Context context) {
-    this.context = context;
+  FeatureHelper(Resources resources) {
+    this.resources = resources;
   }
 
-  public String getCreatedBy(@NonNull Optional<Feature> feature) {
-    return getUserName(feature).map(name -> context.getString(R.string.added_by, name)).orElse("");
+  public String getCreatedBy(Optional<Feature> feature) {
+    return getUserName(feature)
+        .map(name -> resources.getString(R.string.added_by, name))
+        .orElse("");
   }
 
-  public String getTitle(@NonNull Optional<Feature> feature) {
-    return getCaption(feature).orElseGet(() -> getLayerName(feature).orElse(""));
+  public String getLabel(Optional<Feature> feature) {
+    return getCaption(feature).orElseGet(() -> feature.map(this::getFeatureType).orElse(""));
   }
 
-  private Optional<String> getUserName(@NonNull Optional<Feature> feature) {
+  public String getFeatureType(Feature feature) {
+    if (feature.isGeoJson()) {
+      return resources.getString(R.string.polygon);
+    } else {
+      return resources.getString(R.string.point);
+    }
+  }
+
+  public String getSubtitle(Optional<Feature> feature) {
+    return feature
+        .map(f -> resources.getString(R.string.layer_label_format, f.getLayer().getName()))
+        .orElse("");
+  }
+
+  private Optional<String> getUserName(Optional<Feature> feature) {
     return feature.map(Feature::getCreated).map(AuditInfo::getUser).map(User::getDisplayName);
   }
 
-  private Optional<String> getCaption(@NonNull Optional<Feature> feature) {
-    return feature.map(Feature::getCaption).map(String::trim).filter(caption -> !caption.isEmpty());
+  private Optional<String> getCaption(Optional<Feature> feature) {
+    // TODO(#793): Allow user-defined feature names for other feature types.
+    return feature.map(this::getCaption).map(String::trim).filter(caption -> !caption.isEmpty());
+  }
+
+  @Nullable
+  private String getCaption(Feature feature) {
+    if (feature.isGeoJson()) {
+      return getGeoJsonCaption((GeoJsonFeature) feature);
+    }
+    return feature.getCaption();
   }
 
-  private Optional<String> getLayerName(@NonNull Optional<Feature> feature) {
-    return feature.map(Feature::getLayer).map(Layer::getName);
+  private String getGeoJsonCaption(GeoJsonFeature feature) {
+    String caption = feature.getCaptionFromProperties();
+    return caption.isEmpty()
+        ? getFeatureType(feature) + " " + feature.getIdFromProperties()
```

Comment:
optional: Feels like we could declare some of these methods in the interface/abstract class for generic features and let the specific types (e.g. GeoJsonFeature) provide concrete implementations. This would reduce some of the duplication that occurs re: checking the feature type here and above.

Reply:
The trouble is that Feature and related subclasses don't have references to the application context required for i18n. Eventually the model classes will be dumb data model classes (either as Kotlin data classes, or protobufs), in which case the presenters for those classes will need to be separate anyway.

Reply:
Ah, ok, sounds good!